### ISSUE TITLE
cpp_parse_function only works for functions with one argument

### ISSUE THREAD

**Issue by aolo2** [4coder-editor/4coder#118]
In `cpp_parse_function` comma and semicolon are both accepted as the end of the function declaration/definiton, because of this check:

```
if (peek->kind == TokenBaseKind_ParentheticalOpen ||
                     peek->kind == TokenBaseKind_ScopeOpen ||
                     peek->kind == TokenBaseKind_ScopeClose ||
                     peek->kind == TokenBaseKind_StatementClose) {
    break;
}
```

This causes the the code index to ignore functions with > 1 arguments, partially breaking `jump_to_definition`.

The fix is a simple
```
... && peek->sub_kind == TokenCppKind_Semicolon
```

**Comment by aolo2** [comment]
In the same vein, `cpp_parse_function` currently thinks that function definition is over when it sees any of:
```
TokenBaseKind_ScopeOpen,
TokenBaseKind_ScopeClose,
TokenBaseKind_ParentheticalOpen,
TokenBaseKind_ParentheticalClose,
TokenBaseKind_StatementClose
```

However, this doesn't support function pointer arguments, as well as some other stuff, like C++ default values for struct arguments, e.g.:
```
void foo(Bar bar = { 1, 2, 3 })
```

An easier and more correct way of handling function declarations/definitions would be to track the parenthesis balance: the declaration/definition is over when the first opening parenthesis is closed:
```
scope_balance = 1;
for (peek in tokens) {
    if (peek->kind == TokenBaseKind_ParentheticalOpen) { ++scope_balance; }
    if (peek->kind == TokenBaseKind_ParentheticalClose) { --scope_balance; }
    if (scope_balance == 0) { at_paren_close = true; break; }
}
```

**Comment by 4coder-editor** [comment]
Thanks for the report, I used your suggestion for paren balancing and it looks like it's working. The fix is in the patch 4.1.2.
